rir/src/ir/CodeEditor.cpp

```diff
@@ -33,49 +33,50 @@ CodeEditor::CodeEditor(SEXP in) {
         ::Function* f = isValidClosureSEXP(in);
         assert(f != nullptr);
         formals_ = FORMALS(in);
-        DispatchTable* dispatchTable = sexp2dispatchTable(BODY(in));
-        bc = dispatchTable->entry[0];
+        DispatchTable* dispatchTable = DispatchTable::unpack(BODY(in));
+        bc = dispatchTable->first()->container();
     } else {
         assert(isValidFunctionObject(in));
     }
-    FunctionHandle fh(bc);
-    CodeHandle ch = fh.entryPoint();
-    ast = ch.ast();
-    loadCode(fh, ch, true);
+    Function* f = Function::unpack(bc);
+    Code* ch = f->body();
+    ast = src_pool_at(globalContext(), ch->src);
+    loadCode(f, ch, true);
 }
 
-CodeEditor::CodeEditor(CodeHandle code) {
-    ast = code.ast();
-    loadCode(code.function(), code, false);
+CodeEditor::CodeEditor(Code* code) {
+    ast = src_pool_at(globalContext(), code->src);
+    loadCode(code->function(), code, false);
 }
 
-CodeEditor::CodeEditor(CodeHandle code, SEXP formals) {
+CodeEditor::CodeEditor(Code* code, SEXP formals) {
     formals_ = formals;
-    ast = code.ast();
-    loadCode(code.function(), code, true);
+    ast = src_pool_at(globalContext(), code->src);
+    loadCode(code->function(), code, true);
 }
 
-void CodeEditor::loadCode(FunctionHandle function, CodeHandle code, bool loadCompiledDefaultArgs) {
+void CodeEditor::loadCode(Function* function, Code* code,
+                          bool loadCompiledDefaultArgs) {
     std::unordered_map<Opcode*, LabelT> bcLabels;
 
     // Add promises that are default values of formal arguments
     if (loadCompiledDefaultArgs) {
-        for (auto c : function) {
+        unsigned idx = 0;
+        for (auto c : *function) {
             if (c->isDefaultArgument) {
-                CodeHandle ch(c);
-                CodeEditor* p = new CodeEditor(ch);
-                auto idx = ch.idx();
+                CodeEditor* p = new CodeEditor(c);
                 if (promises.size() <= idx)
                     promises.resize(idx + 1, nullptr);
                 promises[idx] = p;
                 defaultArguments.push_back(idx);
             }
+            ++idx;
```

Comment:
This looks weird. `idx` just goes from 0 to \#codes and so the promises vector is always stretched by just 1.

I reused the resizing from the code further down (I think because of the possibility that code handle could skip some indexes?) but here it should probably just be `promises.push_back(p);`

Anyway, also unrelated, so I will make a note and fix it later...

Reply:
hmm, I am not sure what the resize is supposed to do, honestly. but I think that the code does the same as before